Add config changes for UI metrics

banks opened the pull request:
This PR adds new configuration for the UI to support up-coming work on integrating metrics. None of the new configurations do anything yet this is just the wiring for them. Building the features will come in subsequent PRs.

Following the checklist in `/contributing`:

## Adding a Simple Config Field for Client Agents

 - [x] Add the field to the Config struct (or an appropriate sub-struct) in
   `agent/config/config.go`.
 - [x] Add the field to the actual RuntimeConfig struct in
   `agent/config/runtime.go`.
 - [x] Add an appropriate parser/setter in `agent/config/builder.go` to
   translate.
 - [x] Add the new field with a random value to both the JSON and HCL blobs in
   `TestFullConfig` in `agent/config/runtime_test.go`, it should fail now, then
   add the same random value to the expected struct in that test so it passes
   again.
 - [x] Add the new field and it's default value to `TestSanitize` in the same
   file. (Running the test first gives you a nice diff which can save working
   out where etc.)
 - [x] **If** your new config field needed some validation as it's only valid in
   some cases or with some values (often true).
      - [x] Add validation to Validate in `agent/config/builder.go`.
      - [x] Add a test case to the table test `TestConfigFlagsAndEdgeCases` in
        `agent/config/runtime_test.go`.
 - [x] **If** your new config field needs a non-zero-value default.
      - [ ] Add that to `DefaultSource` in `agent/config/defaults.go`.
      - [ ] Add a test case to the table test `TestConfigFlagsAndEdgeCases` in
        `agent/config/runtime_test.go`.
 - [x] **If** your config should take effect on a reload/HUP.
      - [x] Add necessary code to to trigger a safe (locked or atomic) update to
        any state the feature needs changing. This needs to be added to one or
        more of the following places:
         - `ReloadConfig` in `agent/agent.go` if it needs to affect the local
           client state or another client agent component.
         - `ReloadConfig` in `agent/consul/client.go` if it needs to affect
           state for client agent's RPC client.
      - [x] Add a test to `agent/agent_test.go` similar to others with prefix
        `TestAgent_reloadConfig*`.
 - [ ] Add documentation to `website/source/docs/agent/options.html.md`.

## TODO

As can be seen above I've not yet added the documentation for these new configurations. I intend to do that in a later PR once more of the feature is working and we can be more accurate about the actual effect in case any of it doesn't work out as planned!

rboyer commented on agent/http.go:
```diff
@@ -29,6 +29,7 @@ import (
 	"github.com/hashicorp/go-cleanhttp"
 	"github.com/mitchellh/mapstructure"
 	"github.com/pkg/errors"
+	"github.com/ryboe/q"
```
You probably didn't mean to leave this in.

banks replied:
Indeed, I fixed that - the lib changed owner recently so my usual precommit hook didn't catch it. Apparently didn't push the fix yet although I thought I had.

rboyer commented on agent/agent.go:
```diff
@@ -349,6 +359,9 @@ func New(bd BaseDeps) (*Agent, error) {
 	}
 
+	// Initialize the UI Config
+	a.uiConfig.Store(a.config.UIConfig)
```
Did you want to add something like `Clone()` on the UIConfig so you can do a deep copy of these things before stuffing them into a shared atomic? The enclosing struct is not a pointer so the top level will shallow copy, but these fields will not:
```
MetricsProviderFiles       []string
MetricsProxy               UIMetricsProxy    // nested: AddHeaders []UIMetricsProxyAddHeader
DashboardURLTemplates      map[string]string
```

dnephin replied:
Is there a reason to believe we need a copy? Config should generally only be modified by the user and used without modification.

rboyer replied:
Unclear, I just get a little twitchy sometimes with shared concurrent state being handed off like this without seeing all of the ways it might be used (since this PR does not include the use of this stuff, just the config of the stuff).

banks replied:
Yeah I generally assume that nothing will be badly behaved enough to change config structs directly outside of the `config` package.

dnephin commented on agent/agent.go:
```diff
@@ -301,6 +302,15 @@ type Agent struct {
 	// Shared RPC Router
 	router *router.Router
 
+	// uiConfig is an atomically held copy of the current UI config. It's
+	// populated from config.UIConfig on agent startup but then becomes the
+	// canonical source where all reads during runtime should be made. It may be
+	// updated on a configuration reload.
```
It's populated in `Agent.New`, not `Agent.Start`, right? agent startup feels misleading.

*Why* is this field used instead of `a.config` ? I think that is important to capture in the comment.

banks replied:
Yeah I meant "startup" generally - the initial creation of an agent - rather than literally the Start method but I can change the wording if that's confusing.

> Why is this field used instead of a.config ? 

That would be because:

> It may be updated on a configuration reload.

I can make that more explicit though if you didn't read that as a clear enough explanation!

dnephin commented on agent/agent.go:
```diff
@@ -3839,3 +3852,14 @@ func defaultIfEmpty(val, defaultVal string) string {
 	return defaultVal
 }
+
+// getUIConfig is the canonical way to read the value of the UIConfig at
+// runtime. It is thread safe and returns the most recent configuration which
+// may have changed since the agent started due to config reload.
+func (a *Agent) getUIConfig() config.UIConfig {
```
I think we should be avoiding new Agent methods as much as possible. How will this be used? Could this be a method on `HTTPServer` instead? I assume it will be used there?

dnephin replied:
#8623 has a pattern for config reloading that I think might work better. It would remove the need for the extra field and method from `Agent`. It introduces an interface so that each component can reload its config.

banks replied:
Yeah I didn't love putting this here but it seemed cleaner than attaching the atomic UI config state to some other component - I considered putting it on the HTTPServer itself but that felt wrong as the HTTPServer is really just a thin wrapper around the HTTP handler functions and not something that semantically owns the config for the UI.

I like your inverted reload patten in #8623 - I think I would use that once it's landed, however I still feel odd about leaving this config state with no canonically correct way to read it unless you happened to only need it on a reload. Especially since it's already confusing that there are two different places this config can be read from. It seems like it would be more likely to be mis-used (either read from the runtime config instead of the reloaded one or duplicated atomic logic) in the future if there was no canonical accessor on the object that owns the state.

I think overall my ideal approach here would be to make the whole RuntimeConfig atomically reloadable since the config package is really the most semantic place to put logic related to config and reloading, however that's a big enough task not to want to have on critical path for a feature we're trying to ship for now! It requires changing every part of code that accesses the agent config! But it would avoid the confusion you identified below where only parts of the config actually change during a reload and code that refers only to the original config won't ever observe the reload even if it's looking at a "reloadable" value.

banks replied:
Actually, I'm working on the next two components and realise this reloaded UI needs to get referenced in more places and passing a whole agent around gets even more gross than it should be.

I think the pattern you proposed is cleaner and I'll probably switch to that.

dnephin commented on agent/config/builder.go:
```diff
@@ -797,6 +798,26 @@ func (b *Builder) Build() (rt RuntimeConfig, err error) {
+	// Handle Deprecated UI config fields
+	if c.UI != nil {
+		b.warn("The 'ui' field is deprecated. Use the 'ui_config.enabled' field instead.")
+		if c.UIConfig.Enabled == nil {
+			c.UIConfig.Enabled = c.UI
+		}
+	}
```
I think it would be easier to trace how this value is set (and more consistent with other fields) by removing this from here and changing line 1709 below to:

```go
	b.boolValWithDefault(c.UIConfig.Enabled, b.boolVal(c.UI))
```

The warning can be anywhere, it could stay here, but might also be good in `uiConfigVal`.

I think the same is true for the two settings below.

banks replied:
I originally was going to do it that way but ended up not because:

 1. most other `somethingVal` methods only accept the sub-struct from config not the whole thing. In general the pattern is for those to be "happy path" setters/converters from one struct type to another. We'd have to pass the whole Config into the `uiConfigVal` method which is a little different from the prevailing pattern here.
 2. In general other logic which has to move across multiple fields in different parts of the `Config` are handled as special cases here - see the deprecation of acl_datacenter and passing of the value
 3. I can see what you mean about the setting of values all being in one place with your suggestion. This way does have the benefit that all the code for the deprecated fields is in one place and can potentially be removed all as one later. If I were trying to find this i might do so by grepping for the deprecated warning and having the actual defaulting in the same place seems useful rather than having this deprecation handled in multiple places.
 4. The precedent for use of the `WithDefault` values is all within the Builder method directly that has access to the full config - most are in this "prelude" special cases section already with a few that were just mechanical movements (acl_* to acl.*) being in the assignments directly.

dnephin commented on agent/agent_test.go:
```diff
@@ -3490,6 +3490,36 @@ func TestAgent_ReloadConfigTLSConfigFailure(t *testing.T) {
+			metrics_provider = "foo"
+		}
+	`
+	a := NewTestAgent(t, hcl)
+	defer a.Shutdown()
```
This is a very expensive test for what is essentially testing that an atomic works. Do we really need a running agent to be able to test this?

banks replied:
🤔 This is really a test of the plumbing and glue code around reloading. It follows the pattern of the other reload tests. IIRC for a long time we didn't have tests for much of this stuff and often had bugs where code had been added to a reload method but some part of the plumbing missed.

Sadly since the agent isn't more modular it's pretty hard to test any of the plumbing around one without starting one.

I'm with you on wanting to make tests faster and reduce the number of times we spin up a full agent but that seems like something we need to tackle by actually breaking up the agent rather than just not testing new stuff we do add to it?

> Do we really need a running agent to be able to test this?

Maybe not but if we hacked together an agent struct bypassing all the normal startup code and called an internal method to reload a pointer... then it would be only a test of whether the atomic.Value works as expected! If we had a way to start an agent using the normal code paths but skipping some expensive elements like not starting serf or raft maybe that might work?

Overall though, as discussed above I'd rather that this config reload plumbing was all encapsulated in a separate package and not on the agent itself but that's out of scope here.

banks replied:
😄  - this comment made me realise that this test was actually failing...

Because I'd forgotten to actually add the `Store` in the `reloadConfigInternal` method!

Perhaps there is a cheaper way if we change the reload logic a little like you propose in #8623 since then reloaders are not relying on any other state in the agent to be a certain way and we can have a cheaper test that just ensure s all the expected reloaders have indeed been registered after a "normal" startup?

Plumbing tests always seem wasteful since the code is trivial unless it's not there!

dnephin commented on agent/config/runtime.go:
```diff
@@ -1411,16 +1404,18 @@ type RuntimeConfig struct {
-	//UIContentPath is a string that sets the external
-	// path to a string. Default: /ui/
-	UIContentPath string
+	// NOTE: Never read from this field directly once the agent has started up
+	// since the UI config is reloadable. The on in the agent's config field may
+	// be out of date. Use the agent.getUIConfig() method to get the latest config
+	// in a thread-safe way.
+	//
+	// hcl: ui_config { ... }
+	UIConfig UIConfig
```
The runtime reference on the agent is updated on a reload, so wouldn't it be safe to read it from this field as well?

dnephin replied:
Ok, I am wrong about the main config reference. The current `ReloadConfig` is a bit confusing because it generates a new `consul.Config` based on `a.config`, which is not updated... It's not clear to me if `delegate.ReloadConfig` actually does anything. It seems like all it does is pass the same original config to the Client and Server. Maybe I am still missing something.

dnephin replied:
I cleaned this up a bit in #8696

dnephin commented on agent/http_test.go:
```diff
@@ -1198,16 +1198,46 @@ func TestACLResolution(t *testing.T) {
+
+func TestEnableWebUIWithMetricsOptions(t *testing.T) {
+	t.Parallel()
+	a := NewTestAgent(t, `
+		ui_config {
+			enabled = true
+			metrics_provider_options_json = "{\"foo\": 1}"
+		}
+	`)
+	defer a.Shutdown()
```
We don't need a full running agent to test the `HTTPServer`. I think this test could be a lot less expensive.

banks replied:
I agree it could for sure.

Our current testing methodology in this package with a full agent is a know area we'd like to improve but that seems out of scope for a feature PR.

The `HTTPServer` is very limited but most of the methods in it call into the agent state and similar so decoupling the server from the agent state is not trivial. Even in cases like this where we could fake it, I'm not very confident that provides a very robust test. If we had a more holistic refactor - for example make the server only have access to the (reloadable) config from the agent and specific interfaces where it needs access to RPCs and Serf, then we could make tests cheaper without making them brittle but that seems out of scope and fix all the tests not just this one new one!

dnephin reviewed (commented):
A few questions about how these new fields are used, and some suggestions for improving the tests. Our test suite is already quite slow, and if we want to make room for more tests I think we need to be careful about what we are running in each test.

dnephin commented on agent/config/builder.go:
```diff
@@ -1113,12 +1132,50 @@ func (b *Builder) Validate(rt RuntimeConfig) error {
+		// ensures we get an object.
+		var dummyMap map[string]interface{}
+		err := json.Unmarshal([]byte(rt.UIConfig.MetricsProviderOptionsJSON),
+			&dummyMap)
```
Minor: kind of awkward line wrapping (https://github.com/golang/go/wiki/CodeReviewComments#line-length)

banks replied:
I'd tend to agree with you. When I started here it was tribal knowledge that there is a preference in our projects to limit lines to 100 chars max and I was asked to stick to this in several PRs.

I suggested at the time that if this is important we codify it as it's not a general Go community thing but that document  still doesn't exist. I don't know how widespread the practice is but since I setup those line length guides in my editor I still use them.

I'd be happy if the consensus has moved on from that - I know it's not rigorously adhered to anyway - but at one point this was a specific thing people ere seemed to care about the other way 🤷 .

banks replied:
TBF this one was super border line anyway so I just fixed it - obviously more readable on one line. 😄

dnephin commented on agent/config/builder.go:
```diff
@@ -1113,12 +1132,50 @@ func (b *Builder) Validate(rt RuntimeConfig) error {
+			"alphanumeric or _ characters. received: %q", rt.UIConfig.MetricsProvider)
+	}
+	if rt.UIConfig.MetricsProviderOptionsJSON != "" {
+		// Attempt to parse the JSON to ensure it's valid, parsing into a map
+		// ensures we get an object.
+		var dummyMap map[string]interface{}
+		err := json.Unmarshal([]byte(rt.UIConfig.MetricsProviderOptionsJSON),
```
We weren't happy with inlined JSON for connect escape hatches. Is this a similar idea? should we make this opaque config instead and serialize it to the UI?

banks replied:
I considered this during RFC and I'm OK with this as it is for now - it's a highly experimental interface. Somewhat unlike the escape hatches where the JSON is protobuf we decode and merge, this is literally passed through to the front end which passes it directly back to the custom code the operator is using in JavaScript as a parameter. The decode here is just to try to ensure valid formatting so it doesn't just silently break the whole UI if there's a typo.

The UX of writing explicit JSON for a value that is passed to your javascript code doesn't seem as bad to me as writing raw JSON encoded proto3 Envoy config. The complexities of conversions between HCL and JSON seem to potentially make the UX more confusing?

That said, I'm open to iterating on this if it proves its value but I'm not sure the extra complexity of opaque config makes sense for now either from UX or from a "getting an MVP out and seeing if it's useful" perspective?

dnephin commented on agent/config/builder.go:
```diff
@@ -1113,12 +1132,50 @@ func (b *Builder) Validate(rt RuntimeConfig) error {
+	}
+	for k, v := range rt.UIConfig.DashboardURLTemplates {
+		if !reBasicName.MatchString(k) {
+			return fmt.Errorf("ui_config.dashboard_url_templates key names can only "+
+				"contain lowercase alphanumeric or _ characters. received: %q", k)
+		}
```
This is repeated a couple times. Instead of sharing just the regex, possibly expose the functionality with a `validateBasicName(...)` so that others can re-use it without the same duplication?

banks commented:
Thanks @rboyer and @dnephin for the reviews!

I think I fixed up a few things including one of the new tests actually failing!

I've replied inline to a few comments - I think I agree with all of them at some level but several I'm not sure I see a good solution for in this PR without blowing out the scope. Happy to chat more about that though! I've generally followed an existing convention for most parts of this PR, plenty of room for improvement overall but I think to do so substantially would move into major refactor territory.

banks commented:
@dnephin I've refactored this a bunch following your lead on `ReloadConfig`. I think it's much better although I hit two plumbing issues that are not ideal but I couldn't afford to do a lot more refactoring to fix just now:

 1. I had to memoize `HTTPHandlers.handlers` since some components setup there are now stateful (i.e. have reloadable state) which means that having every test run against a different instance wasn't really OK.

 2. On the same lines, it took me a long time to track down why agent package reload tests were failing even after fixing 1 above though the uiserver reload works fine. It turns out that as part of #8234 you moved `TestAgent.srv` to be a different instance rather than one of the actual instances of the agent's Server/handlers. That makes total sense but is broken now that the `HTTPHandlers` is somewhat stateful. Just plumbing that separate instance of `HTTPHandlers` into reloading in `TestAgent` would be possible but then it would be impossible to verify in a test that the actual agent does in fact reload http server state correctly as part of the reloading code paths would be only in test agent code. I opted for now for an unpleasant option - re-introducing an explicit reference to (one of) the `HTTPHandlers`. I think this is least bad for now and can go away when we do actually get around to moving `apiServers` to be an injected dep of `Agent` since the test agent will be able to inject it's own instance then. Does that seem good to you? I'm keen not to scope creep this refactor any more as it's on the critical path for some features that should be complete in O(week).

mikemorris commented on agent/uiserver/uiserver_test.go:
```diff
@@ -0,0 +1,230 @@
+		{
+			// TODO: is this really what we want? It's what we've always done but
+			// seems a bit odd to not do an actual 301 but instead serve the
+			// index.html from every path... It also breaks the UI probably.
```
I'm guessing this is intentional because the UI is a single-page app that parses the path to decide what to render, like when navigating to `http://localhost:8500/ui/dc1/services` we don't actually want to redirect, but rather serve the UI code and let JavaScript control rendering the `dc1` services "page".

banks replied:
Ah good point. I'll remove the TODO.

dnephin reviewed (commented):
LGTM, just one question about the error handling

banks commented on agent/uiserver/uiserver.go:
```diff
@@ -0,0 +1,174 @@
+	h := &Handler{
+		logger: logger.Named(logging.UIServer),
+	}
+	// Don't return the error since this is likely the result of a
+	// misconfiguration and reloading config could fix it. Instead we'll capture
+	// it and return an error for all calls to ServeHTTP so the misconfiguration
+	// is visible. Sadly we can't log effectively
+	h.err = h.ReloadConfig(agentCfg)
```
Good spot, thanks. Yes if this fails it becomes fatal currently. I'll fix that.

banks replied:
Fixed in next commit.

rboyer commented on agent/http_oss.go:
```diff
@@ -55,8 +55,6 @@ func (s *HTTPHandlers) rewordUnknownEnterpriseFieldError(err error) error {
 	return err
 }
 
-func (s *HTTPHandlers) addEnterpriseUIENVVars(_ map[string]interface{}) {}
```
Reminder to clean up this line's counterpart.

rboyer commented on agent/http_test.go:
```diff
@@ -417,62 +418,59 @@ func TestHTTPAPI_TranslateAddrHeader(t *testing.T) {
-	if xss != "1; mode=block" {
-		t.Fatalf("bad X-XSS-Protection header: expected %q, got %q", "1; mode=block", xss)
-	}
+	require.Equal(t, "*", hdrs.Get("Access-Control-Allow-Origin"),
```
Why is Access-Control-Allow-Origin checked twice?

banks replied:
🍝

rboyer commented on agent/uiserver/bindata_assetfs.go:
```diff
@@ -15,7 +15,7 @@
 // pkg/web_ui/assets/codemirror/mode/ruby/ruby-ea43ca3a3bdd63a52811e8464d66134b.js
```
You may need to also edit the CI configs in oss or ent, as this file is specially handled in merges.

rboyer commented on agent/uiserver/buffered_file.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+func (t *bufferedFile) Close() error {
+	// Reset the read cursor since the next "open" will assume it is reading from
```
Did you mean to actually Seek(0,0) here?

banks replied:
Heh I did when i wrote that comment, and then I made it work a different way be refactoring this to wrap a `bytes.Reader` instead of a `bytes.Buffer` so each "instance" had it's own cursor state. I've removed the comment in next commit! Good 👀 .

banks commented:
@dnephin @rboyer Thanks again for the reviews.

I've fixed up those last few issues noticed and rebased on master to resolve a few conflicts.

I've just spotted a legit test failure and lint issue though so I'll fix those up ready to merge.

banks commented:
OK... I think this is good to go. The test fail was legitimately a regression although quite an edge case (when using a custom UI dir that didn't have an index.html (which would be somewhat broken anyway because we redirect index requests to index.html).

Still this is handled slightly more nicely now - we still serve the custom UI and disable attempts to redirect to a non-existent /index.html. I also added a warning log to make the UX a little nicer.

rboyer commented on agent/uiserver/uiserver.go:
```diff
@@ -0,0 +1,190 @@
+	// have to match the encoded double quotes around the JSON string value that
+	// is there as a placeholder so the end result is an actual JSON bool not a
+	// string containing "false" etc.
+	re := regexp.MustCompile(`%22__RUNTIME_BOOL_[A-Za-z0-9-_]+__%22`)
```
This could be a package constant.

banks replied:
Since it's not on a hot path (typically called once per agent lifetime) and very unlikely to be useful in any other place do you think it's preferable to define it out of context on the package? I don't mind but my feeling was it's easier to read this code with it inline so you can see what it's doing all at once.

rboyer reviewed (approved):
LGTM.

You still have a few dangling references to the old `agent/bindata_assetfs.go` name to fixup. The rest of my comments were not blocking.
